src/backend/cuda/nvrtc/cache.hpp

```diff
@@ -0,0 +1,166 @@
+/*******************************************************
+ * Copyright (c) 2019, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#pragma once
+
+#include <err_cuda.hpp>
+#include <nvrtc/launch_args.hpp>
+#include <traits.hpp>
+
+#include <cstdio>
+#include <string>
+#include <vector>
+
+#define CU_CHECK(fn)                                                      \
+    do {                                                                  \
+        CUresult res = fn;                                                \
+        if (res == CUDA_SUCCESS) break;                                   \
+        char cu_err_msg[1024];                                            \
+        const char *cu_err_name;                                          \
+        const char *cu_err_string;                                        \
+        cuGetErrorName(res, &cu_err_name);                                \
+        cuGetErrorString(res, &cu_err_string);                            \
+        snprintf(cu_err_msg, sizeof(cu_err_msg), "CU Error %s(%d): %s\n", \
+                 cu_err_name, (int)(res), cu_err_string);                 \
+        AF_ERROR(cu_err_msg, AF_ERR_INTERNAL);                            \
+    } while (0)
+
+namespace cuda {
+
+///
+/// \brief Kernel Functor that wraps CUDA nvrtc constructs
+///
+/// This struct encapsulates CUmodule and CUfunction pointers that are required
+/// to execution of CUDA C++ kernels compiled at runtime.
+///
+struct Kernel {
+    CUmodule prog;  ///< CUmodule helps acquire kernel attributes
+    CUfunction ker; ///< CUfuntion is the actual kernel blob to run
+
+    ///
+    /// \brief Copy data to constant qualified global variable of kernel
+    ///
+    /// This function copies data of `bytes` size from the device pointer to a
+    /// global(__constant__) variable declared inside the kernel.
+    ///
+    /// \param[in] name is the name of the global variable inside kernel
+    /// \param[in] src is the device pointer from which data will be copied
+    /// \param[in] bytes are the number of bytes of data to be copied
+    ///
+    void setConstant(const char* name, CUdeviceptr src, size_t bytes);
+
+    ///
+    /// \brief operator() overload to execute kernel
```

Comment:
The "operator() overload" is redundant. A better brief would be

Calls a kernel given a set of Enqueue Args and kernel parameters.

Even that seems meh.